sound/soc/sof/core.c

```diff
@@ -365,6 +365,7 @@ static int sof_init_environment(struct snd_sof_dev *sdev)
 err_machine_check:
 	if (ret) {
 		snd_sof_remove(sdev);
+		snd_sof_remove_late(sdev);
```

Comment:
I don't really buy this. 

When this branch is reached, we've already done the probe_early() and probe().
It would seem reasonable that the remove() and remove_late() are called here for symmetry.

@ujfalusi @kv2019i what do you think?

Reply:
and what happens if you remove the modules at this point? Would that trigger other issue?

Reply:
Thank you for your comments!

> I don't really buy this.
> 
> When this branch is reached, we've already done the probe_early() and probe(). It would seem reasonable that the remove() and remove_late() are called here for symmetry.

@plbossart I am not sure If I understood you correctly. From my experiments, it looks like there is an asymmetry in probe_early() and remote_late() which cause the extra pm_runtime reference being held after sof_probe_work fails. Hence, the change made sense to me. However, there could be a better place to add this. (can't claim to be an expert here!)

>and what happens if you remove the modules at this point? Would that trigger other issue?

@ujfalusi I have not tried any such scenarios. I can give it a try. Let me know if you would like me to try anything else.

Note: I just sent out the patch thinking it could be a minor miss. If the problem needs a more complex solution, feel free to post it.

Reply:
Come to think of it, there might be other exit points which might need remote_late(). I will have a relook.

Reply:
@plbossart The change makes sense to me. As you said probe_early() and probe() are done at that point, so, we need to call remove() and remove_late() for symmetry. In fact, I think we need to add remove_late() in every exit points of sof_init_environment() as probe_early() is called before sof_init_environment().

Reply:
I have now pushed a second version taking care of the other exit points. Kindly check.